AMDGPU/GlobalISel: lane masks merging

petar-avramovic opened the pull request:
Basic implementation of lane mask merging for GlobalISel.
Lane masks on GlobalISel are registers with sgpr register class
and S1 LLT - required by machine uniformity analysis.
Implements equivalent of lowerPhis from SILowerI1Copies.cpp in:
patch 1: https://github.com/llvm/llvm-project/pull/75340
patch 2: https://github.com/llvm/llvm-project/pull/75349
patch 3: https://github.com/llvm/llvm-project/pull/80003
patch 4: https://github.com/llvm/llvm-project/pull/78431
patch 5: is in this commit:

AMDGPU/GlobalISelDivergenceLowering: constrain incoming registers

Previously, in PHIs that represent lane masks, incoming registers
taken as-is were not selected as lane masks. Such registers are not
being merged with another lane mask and most often only have S1 LLT.
Implement constrainAsLaneMask by constraining incoming registers
taken as-is with lane mask attributes, essentially transforming them
to lane masks. This is final step in having PHI instructions created
in this pass to be fully instruction-selected.

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,66 @@
+#include "AMDGPU.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+
+#define DEBUG_TYPE "global-isel-divergence-lowering"
```
Do you have better suggestions for pass name?

Pierre-vh replied:
`amdgpu-gi-divergence-lowering` to keep it short?

arsenm commented on llvm/include/llvm/CodeGen/MachineSSAUpdater.h:
```diff
@@ -40,8 +40,8 @@ class MachineSSAUpdater {
-  /// VRC - Register class of the current virtual register.
-  const TargetRegisterClass *VRC = nullptr;
+  /// RegAttrs - current virtual register, new registers copy its attributes.
+  Register RegAttrs;
```
One downside is a TargetRegisterClass is truly supposed to be constant. The register's attributes could change in the background

petar-avramovic replied:
Do you have suggestion how to implement this part, I wanted to avoid virtual methods

arsenm commented on llvm/lib/CodeGen/MachineSSAUpdater.cpp:
```diff
@@ -158,9 +153,9 @@ Register MachineSSAUpdater::GetValueInMiddleOfBlock(MachineBasicBlock *BB,
-                                        BB, BB->getFirstTerminator(),
-                                        VRC, MRI, TII);
+    MachineInstr *NewDef =
+        InsertNewDef(TargetOpcode::IMPLICIT_DEF, BB, BB->getFirstTerminator(),
```
Should really use G_IMPLICIT_DEF if it's a generic virtual register

petar-avramovic replied:
Lane mask registers created in this pass are not generic, they have some sgpr register class and must also have S1 LLT. This is requirement for machine uniformity info to work correctly correctly.

arsenm commented on llvm/lib/CodeGen/MachineSSAUpdater.cpp:
```diff
@@ -197,8 +192,8 @@ Register MachineSSAUpdater::GetValueInMiddleOfBlock(MachineBasicBlock *BB,
-  MachineInstrBuilder InsertedPHI = InsertNewDef(TargetOpcode::PHI, BB,
-                                                 Loc, VRC, MRI, TII);
+  MachineInstrBuilder InsertedPHI =
+      InsertNewDef(TargetOpcode::PHI, BB, Loc, RegAttrs, MRI, TII);
```
Same as with IMPLICIT_DEF, G_PHI for generic virtual registers

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -207,7 +207,39 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
+// PHI where all register operands are sgpr(register class) with S1 LLT.
+bool isLaneMaskPhi(MachineInstr &I, MachineRegisterInfo *MRI,
+                   const SIRegisterInfo &TRI) {
+  if (I.getOpcode() != AMDGPU::PHI)
```
This was meant to verify that lane mask PHIs are correctly selected if they were selected before instruction-selection. Idea was to not split selection of lane mask PHIs across multiple passes

nhaehnle replied:
So the point is that if we see an `PHI` here (as opposed to a `G_PHI`), that's because it has already been selected.

Having this as a separate helper is confusing IMHO. Better to inline a check against `PHI` into `selectPHI`, and add an `assert` there that the defined register is a lane mask in that case.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  StringRef getPassName() const override {
+    return "GlobalISel divergence lowering";
```
Needs `AMDGPU` prefix too

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// GlobalISel pass that selects divergent i1 phis as lane mask phis.
+// Lane mask merging uses same algorithm as SDAG in SILowerI1Copies.
+// Handles all cases of temporal divergence.
+//
+// For divergent non-phi i1 and uniform i1 uses outside of the cycle this pass
+// currently depends on LCSSA to insert phis with one incoming.
```
really small nit: Use three slashes + start with `\file`, first line should be `/// \file`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+    return;
+
+  MRI->setRegClass(DstReg, ST->getBoolRC());
+  return;
```
extra `return`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+
+void DivergenceLoweringHelper::collectIncomingValuesFromPhi(
+    const MachineInstr *MI, SmallVectorImpl<Incoming> &Incomings) const {
+  for (unsigned i = 1; i < MI->getNumOperands(); i += 2) {
```
`i` -> `I` (uppercase)

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+}
+
+// Get pointers to build instruction just after MI (skips phis if needed)
+static std::pair<MachineBasicBlock *, MachineBasicBlock::iterator>
```
`static` is implied inside anonymous namespaces

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+static std::pair<MachineBasicBlock *, MachineBasicBlock::iterator>
+getInsertAfterPtrs(MachineInstr *MI) {
+  MachineBasicBlock *InsertMBB = MI->getParent();
+  return std::make_pair(
```
small nit: i think `make_pair ` isn't needed, you can just use `{A, B}`
Also perhaps put `InsertMBB->SkipPHIsAndLabels(std::next(MI->getIterator())` in a variable, or the `std::next` part at least, to make it less dense

petar-avramovic commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.h:
```diff
@@ -0,0 +1,79 @@
+public:
+  PhiLoweringHelper(MachineFunction *MF, MachineDominatorTree *DT,
+                    MachinePostDominatorTree *PDT);
+  virtual ~PhiLoweringHelper() = default;
```
There are two:
class Vreg1LoweringHelper : public PhiLoweringHelper in SILowerI1Copies.cpp for sdag 
class DivergenceLoweringHelper : public PhiLoweringHelper in AMDGPUGlobalISelDivergenceLowering.cpp for globalisel

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+} // End anonymous namespace.
+
+INITIALIZE_PASS_BEGIN(AMDGPUGlobalISelDivergenceLowering, DEBUG_TYPE,
+                      "GlobalISel divergence lowering", false, false)
```
`AMDGPU` prefix here and in `_END`

nhaehnle replied:
This comment was addressed in the wrong commit. This makes it harder to review the PR :(

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+      &MF, &DomTree, &getAnalysis<MachinePostDominatorTree>(), &MUI);
+
+  Helper.lowerPhis();
+  return true;
```
Doesn't track changes, if the function didn't change we might end up doing extra work for nothing (I think it's how it works?)

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -207,7 +207,39 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
+  if (MRI->getType(Reg) != LLT::scalar(1))
+    return false;
+  const TargetRegisterClass *RC = MRI->getRegClassOrNull(Reg);
+  if (!RC || !TRI.isSGPRClass(RC))
+    return false;
+
+  return true;
```
`return RC && TRI.isSGPRClass(RC);`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+    MachineBasicBlock &MBB, MachineBasicBlock::iterator I, const DebugLoc &DL,
+    Register DstReg, Register PrevReg, Register CurReg) {
+  // TODO: check if inputs are constants or results of a compare.
+
```
nit: could use some docs. In this type of function I like adding some pseudo-MIR example of what it emits

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+    }
+  }
+
+  return;
```
extra `return`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+  MRI->setRegClass(Copy.getReg(0), ST->getBoolRC());
+  In.Reg = Copy.getReg(0);
+
+  return;
```
extra `return`

nhaehnle replied:
This extra `return` in `constrainIncomingRegisterTakenAsIs` should still be removed.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -0,0 +1,213 @@
+      getAnalysis<MachineCycleInfoWrapperPass>().getCycleInfo();
+  MachineDominatorTree &DomTree = getAnalysis<MachineDominatorTree>();
+
+  MachineUniformityInfo MUI =
```
Can't you depend on `MachineUniformityAnalysis` instead?

Pierre-vh commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.h:
```diff
@@ -0,0 +1,79 @@
+#include "llvm/CodeGen/MachinePostDominators.h"
+#include "llvm/CodeGen/MachineSSAUpdater.h"
+
+using namespace llvm;
```
Avoid `using namespace` in headers

Pierre-vh commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.h:
```diff
@@ -0,0 +1,79 @@
+
+using namespace llvm;
+
+struct Incoming {
```
docs?

Pierre-vh reviewed (commented):
Some more docs in `AMDGPUGlobalISelDivergenceLowering.cpp` would be nice.
Seems like it's going to be a core part of the GlobalISel pipeline so IMO it should be well documented so it's not too hard to understand and contribute to.

petar-avramovic committed:
AMDGPU/GlobalISel: lane masks merging (#73337)

Basic implementation of lane mask merging for GlobalISel.
Lane masks on GlobalISel are registers with sgpr register class
and S1 LLT - required by machine uniformity analysis.
Implements equivalent of lowerPhis from SILowerI1Copies.cpp in:
patch 1: https://github.com/llvm/llvm-project/pull/75340
patch 2: https://github.com/llvm/llvm-project/pull/75349
patch 3: https://github.com/llvm/llvm-project/pull/80003
patch 4: https://github.com/llvm/llvm-project/pull/78431
patch 5: is in this commit:

AMDGPU/GlobalISelDivergenceLowering: constrain incoming registers

Previously, in PHIs that represent lane masks, incoming registers
taken as-is were not selected as lane masks. Such registers are not
being merged with another lane mask and most often only have S1 LLT.
Implement constrainAsLaneMask by constraining incoming registers
taken as-is with lane mask attributes, essentially transforming them
to lane masks. This is final step in having PHI instructions created
in this pass to be fully instruction-selected.

petar-avramovic commented:
Ping. Updated for comments about code quality. Unchanged for comments about functional changes:
Divergent i1 phis (G_PHIs to be precise) are fully instruction selected into PHIs (all registers created in this pass are lane masks with correct register class and S1 LLT). Check regression tests changes (last two patches in particular) to get better idea what is happening.

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/GlobalISel/divergence-structurizer.mir:
```diff
@@ -29,18 +30,18 @@ body: |
   ; GFX10-NEXT: bb.2:
-  ; GFX10-NEXT:   [[PHI:%[0-9]+]]:sreg_32(s1) = PHI [[ICMP]](s1), %bb.0, [[S_OR_B32_]](s1), %bb.1
-  ; GFX10-NEXT:   [[COPY6:%[0-9]+]]:sreg_32(s1) = COPY [[PHI]](s1)
+  ; GFX10-NEXT:   [[PHI:%[0-9]+]]:sreg_32(s1) = PHI [[COPY4]](s1), %bb.0, [[S_OR_B32_]](s1), %bb.1
```
> What's the point of this function? Don't really understand the name constrainIncomingRegisterTakenAsIs

But here register operand with name COPY4 has lane-mask register class and after this patch lane mask PHI is fully selected. I wanted to avoid having to split selection of lane mask phis across multiple passes.
  ; GFX10-NEXT:   [[ICMP:%[0-9]+]]:_(s1) = G_ICMP intpred(uge), [[COPY2]](s32), [[C]]
  ; GFX10-NEXT:   [[COPY4:%[0-9]+]]:sreg_32(s1) = COPY [[ICMP]](s1)
  ; GFX10-NEXT:   [[PHI:%[0-9]+]]:sreg_32(s1) = PHI [[COPY4]](s1), %bb.0, [[S_OR_B32_]](s1), %bb.1

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/GlobalISel/divergence-structurizer.mir:
```diff
@@ -29,18 +30,18 @@ body: |
   ; GFX10-NEXT:   [[S_OR_B32_:%[0-9]+]]:sreg_32(s1) = S_OR_B32 [[S_ANDN2_B32_]](s1), [[S_AND_B32_]](s1), implicit-def $scc
   ; GFX10-NEXT: {{  $}}
   ; GFX10-NEXT: bb.2:
-  ; GFX10-NEXT:   [[PHI:%[0-9]+]]:sreg_32(s1) = PHI [[ICMP]](s1), %bb.0, [[S_OR_B32_]](s1), %bb.1
```
> What's the point of this function? Don't really understand the name constrainIncomingRegisterTakenAsIs

That Incoming is from the block that is visited first in RPOT, so it is taken as is. But it does not have reg class (the register operand with name ICMP), so PHI instruction is only partially selected

  ; GFX10-NEXT:   [[ICMP:%[0-9]+]]:_(s1) = G_ICMP intpred(uge), [[COPY2]](s32), [[C]]
  ; GFX10-NEXT:   [[PHI:%[0-9]+]]:sreg_32(s1) = PHI [[ICMP]](s1), %bb.0, [[S_OR_B32_]](s1), %bb.1

petar-avramovic commented:
Ping. Nicolai and Jay do you have comments.

nhaehnle reviewed (commented):
It's probably for the best to land this PR soon. I have a few detail comments still inline, but hopefully they're not too complex.

I am somewhat surprised about the order in which the changes appear. Though you did update the tests, so did you actually run a check (e.g. using `git rebase -x "ninja -C build check-llvm-codegen-amdgpu" ...`) that passes after each individual commit? If yes, then great. That would make me feel a bit more confident about the structure of this.

petar-avramovic commented:
> I am somewhat surprised about the order in which the changes appear. Though you did update the tests, so did you actually run a check (e.g. using `git rebase -x "ninja -C build check-llvm-codegen-amdgpu" ...`) that passes after each individual commit? If yes, then great. That would make me feel a bit more confident about the structure of this.

Each individual patch passes check-llvm, ir is correct after the last patch but mir updates are relevant and give good overview of the changes (PHIs go from partially selected to fully selected). I wanted to open separate pr for each of the patches if they are ok together (don't know if there is a better way).

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -207,9 +207,31 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
   const LLT DefTy = MRI->getType(DefReg);
+  // Lane mask PHIs, PHI where all register operands have sgpr register class
+  // with S1 LLT, are already selected in divergence lowering pass.
+  if (I.getOpcode() == AMDGPU::PHI && isLaneMask(DefReg, MRI, TRI)) {
```
Perhaps I should be clearer. I don't think we ever have a `PHI` here other than for lane masks. So it would be better for understanding the overall context to say something like:
```
  if (I.getOpcode() == AMDGPU::PHI) {
    assert(isLaneMask(DefReg, MRI, TRI);
    return true;
  }
```

petar-avramovic replied:
There is one regression test with PHI in `phi_s32_ss_sbranch - GlobalISel/inst-select-phi.mir `. It can be disabled. Other then that assert is fine.
I wanted to check isLaneMask for all operand, can I bring helper function back then? it would look like this without helper
```
  if (I.getOpcode() == AMDGPU::PHI) {
    assert(isLaneMask(DefReg, MRI, TRI));
    for (unsigned i = 1, e = I.getNumOperands(); i != e; e += 2)
      assert(isLaneMask(I.getOperand(i).getReg(), MRI, TRI));
    return true;
  }
```

nhaehnle replied:
Huh, I wonder if @arsenm knows where that test comes from, whether that's on purpose or perhaps a mistake and the test should be removed.

If my understanding is wrong and other PHIs are possible, it would be good to understand why. If they aren't, then the test should be removed and it's not so important to check all the operands IMHO.

petar-avramovic replied:
Okay then, we can merge first two patches in the meantime

nhaehnle reviewed (approved):
Thanks, LGTM except for one nitty nit.

ruiling commented on llvm/test/CodeGen/AMDGPU/GlobalISel/divergence-temporal-divergent-i1.ll:
```diff
@@ -96,30 +102,41 @@ define amdgpu_cs void @loop_with_1break(ptr addrspace(1) %x, i32 %x.size, ptr ad
-; GFX10-NEXT:    s_or_b32 s4, s0, s4
-; GFX10-NEXT:    s_and_b32 s0, 1, s1
-; GFX10-NEXT:    v_cmp_ne_u32_e64 s0, 0, s0
+; GFX10-NEXT:    s_and_b32 s5, 1, s5
+; GFX10-NEXT:    s_and_b32 s6, exec_lo, s1
+; GFX10-NEXT:    v_cmp_ne_u32_e64 s5, 0, s5
+; GFX10-NEXT:    s_or_b32 s4, s6, s4
+; GFX10-NEXT:    s_andn2_b32 s0, s0, exec_lo
+; GFX10-NEXT:    s_and_b32 s5, exec_lo, s5
+; GFX10-NEXT:    s_or_b32 s0, s0, s5
```
Why does the change regress code generation here?

petar-avramovic replied:
See the run line, previous patch disabled ll test because it would crash (some registers in PHI did not have selected register class)
With this patch we are able to generate some assembler output so it includes lane mask merging from previous patch. Code size regression is required for correctness. There is room for improvements in code size but that will follow up later.
Check the mir version of the same test, you will see what changed in this patch specifically

petar-avramovic commented on llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-phi.mir:
```diff
@@ -422,7 +422,7 @@ body:             |
 
   bb.2:
-    %6:sgpr(s32) = PHI %0(s32), %bb.0, %5(s32), %bb.1
+    %6:sgpr(s32) = G_PHI %0(s32), %bb.0, %5(s32), %bb.1
```
Changed to G_PHI

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -210,6 +210,14 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
+  // Lane mask PHIs, PHI where all register operands have sgpr register class
+  // with S1 LLT, are already selected in divergence lowering pass.
+  if (I.getOpcode() == AMDGPU::PHI) {
+    assert(MRI->getType(DefReg) == LLT::scalar(1));
+    assert(TRI.isSGPRClass(MRI->getRegClass(DefReg)));
```
Changed to assert on DefReg only for PHIs is @arsenm agrees.

arsenm commented:
Needs update to resolve conflict

petar-avramovic commented:
Updated with final patch from original PR

arsenm reviewed (commented):
Can the amdgpu-global-isel-risky-select flag be deleted?

petar-avramovic commented:
> Can the amdgpu-global-isel-risky-select flag be deleted?

Yes, will do it in another patch, from the upcoming patches:
`S1 G_PHIs` should not be selected in instruction-select, instead:
- divergent `S1 G_PHI` is pre-selected in AMDGPUGlobalISelDivergenceLowering (into fully inst-selected PHI) 
- uniform `S1 G_PHI` is lowered into `S32 G_PHI` in AMDGPURegBankSelect.cpp (TODO)
